src/main/java/org/primefaces/component/api/UIData.java

```diff
@@ -259,7 +267,10 @@ public void setRowStatePreserved(boolean _paginator) {
     }
 
     public void resetRows() {
-        getStateHelper().remove(PropertyKeys.rows);
+        Object rows = getStateHelper().eval(PropertyKeys.internal_rowsInitialValue);
+        if (rows != null) {
+            setRows((int) rows);
+        }
```

Comment:
Need some more time to think about this, because we should not break https://github.com/primefaces/primefaces/issues/4573.

Reply:
IMO resetRows as it is up to this PR does only work when rows is bound to a value expression.
It does not work when the attribute-value is written direct into the xhtml. (rows="5")